gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -325,11 +358,34 @@ private void redirectToAuthorize(RoutingContext routingContext, Client client) {
         }
     }
 
-    private void verify(FactorProvider factorProvider, FactorContext factorContext, Handler<AsyncResult<Void>> handler) {
+    private void verify(FactorProvider factorProvider, FactorContext factorContext, Optional<VerifyAttempt> verifyAttempt,
+                        RoutingContext routingContext, Handler<AsyncResult<Void>> handler) {
         factorProvider.verify(factorContext)
                 .subscribe(
-                        () -> handler.handle(Future.succeededFuture()),
-                        error -> handler.handle(Future.failedFuture(error)));
+                        () -> {
+                            if (verifyAttempt.isPresent()) {
+                                verifyAttemptService.delete(verifyAttempt.get().getId()).subscribe(
+                                        () -> handler.handle(Future.succeededFuture()),
+                                        error -> {
+                                            logger.error("Could not delete verify attempt", error);
+                                            routingContext.fail(401);
```

Comment:
Do we have to reject the authentication/code validation if the delete action fails ?

Reply:
no, we do not as I understand. Delete performs after successful verification. That means in this stage everything is fine (challenge verified) and deletes if there is any existing verification attempts in the db. It shouldn't have any impact in verification process if the delete is failed for unexpected reason in my opinion.